File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isLetter, isSpace)
import Data.List (dropWhileEnd)

responseFor :: String -> String
responseFor xs
    | null trimmed            = "Fine. Be that way!"
    | isQuestion && isShouting = "Calm down, I know what I'm doing!"
    | isShouting               = "Whoa, chill out!"
    | isQuestion               = "Sure."
    | otherwise                = "Whatever."
  where
    -- Trim leading and trailing whitespace
    trimmed     = dropWhileEnd isSpace (dropWhile isSpace xs)
    letters     = filter isLetter trimmed
    isQuestion  = not (null trimmed) && last trimmed == '?'
    isShouting  = not (null letters) && all isUpper letters

File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    trimmed = String.trim(to_string(input))

    cond do
      trimmed == "" ->
        "Fine. Be that way!"

      all_caps?(trimmed) and String.ends_with?(trimmed, "?") ->
        "Calm down, I know what I'm doing!"

      all_caps?(trimmed) ->
        "Whoa, chill out!"

      String.ends_with?(trimmed, "?") ->
        "Sure."

      true ->
        "Whatever."
    end
  end

  defp all_caps?(line) do
    letters = Regex.scan(~r/\p{L}/u, line) |> List.flatten() |> Enum.join()

    letters != "" and String.upcase(letters) == letters
  end
end
